### PR TITLE
CLOUDP-240782: Fix Serverless Private Endpoint

### PR DESCRIPTION
**Pull request by helderjs**
### All Submissions:

* [x] Have you signed our [CLA](https://www.mongodb.com/legal/contributor-agreement)?
* [ ] Put `closes #XXXX` in your comment to auto-close the issue that your PR fixes (if there is one).
* [ ] Update docs/release-notes/release-notes.md if your changes should be included in the release notes for the next release.

While focused on fixing serverless private endpoint behavior, this PR accomplishes several goals:

* Fix any hidden bugs that may exist in the previous super complex implementation (I know I might be inserting new ones)
* Make implementation clear and easy to understand
* Cover with (a lot) of tests
* Use the new SDK (once the code was completely reworked)

### CODE REVIEW

**Review by josvazg — APPROVED** [review]
LGTM, great work!

**Review by igor-karpukhin — APPROVED** [review]
Good job! 👍

File: pkg/controller/atlasdeployment/serverless_private_endpoint.go [hunk 1]
```diff
@@ -1,459 +1,271 @@
...
-	SPEStatusAvailable  = "AVAILABLE"  //stage 2
-	SPEStatusInitiating = "INITIATING" //stage 2
-	SPEStatusFailed     = "FAILED"     //stage 2
+	SPEStatusDeleting  = "DELETING"
+	SPEStatusReserved  = "RESERVED"
+	SPEStatusAvailable = "AVAILABLE"
 )
 
-func ensureServerlessPrivateEndpoints(service *workflow.Context, groupID string, deployment *akov2.AtlasDeployment, deploymentName string, protected bool) workflow.Result {
+func ensureServerlessPrivateEndpoints(service *workflow.Context, projectID string, deployment *akov2.AtlasDeployment) workflow.Result {
```
Comment (s-urbaniak) [inline pkg/controller/atlasdeployment/serverless_private_endpoint.go:23]: This function can be simplified to:
```
func ensureServerlessPrivateEndpoints(service *workflow.Context, projectID string, deployment *akov2.AtlasDeployment) workflow.Result {
	if deployment == nil || deployment.Spec.ServerlessSpec == nil {
		return workflow.Terminate(workflow.Internal, "serverless deployment spec is empty")
	}

	deploymentSpec := deployment.Spec.ServerlessSpec

	if isGCPWithPrivateEndpoints(deploymentSpec) {
		return workflow.Terminate(workflow.AtlasUnsupportedFeature, "serverless private endpoints are not supported for GCP")
	}

	if isGCPWithoutPrivateEndpoints(deploymentSpec) {
		return workflow.OK()
	}

	finished, err := syncServerlessPrivateEndpoints(service, projectID, deploymentSpec)
	if err != nil {
		return workflow.Terminate(workflow.ServerlessPrivateEndpointFailed, err.Error())
	}

	result := workflow.OK()
	if !finished {
		result = workflow.InProgress(workflow.ServerlessPrivateEndpointInProgress, "Waiting serverless private endpoint to be configured")
	}

	switch len(deploymentSpec.PrivateEndpoints) {
	case 0:
		service.UnsetCondition(status.ServerlessPrivateEndpointReadyType)
	default:
		service.SetConditionFromResult(status.ServerlessPrivateEndpointReadyType, result)
	}

	return result
}
```

**Review by s-urbaniak — COMMENTED** [review]
This looks great, thank you! 🎉 A couple of additional comments, please address them as a follow-up PR.

File: pkg/controller/atlasdeployment/serverless_private_endpoint.go [hunk 2]
```diff
@@ -1,459 +1,271 @@
...
+	toUpdate := make([]akov2.ServerlessPrivateEndpoint, 0, len(privateEndpoints))
+	toDelete := make([]admin.ServerlessTenantEndpoint, 0, len(atlasPrivateEndpoints))
+
+	privateEndpointsByName := map[string]*akov2.ServerlessPrivateEndpoint{}
+	for i := range privateEndpoints {
+		privateEndpoint := privateEndpoints[i]
+		privateEndpointsByName[privateEndpoint.Name] = &privateEndpoint
+	}
+
+	atlasPrivateEndpointsByName := map[string]*admin.ServerlessTenantEndpoint{}
```
Comment (s-urbaniak) [inline pkg/controller/atlasdeployment/serverless_private_endpoint.go:205]: this should technically be called `atlasPrivateEndpointsByComment`

Comment (s-urbaniak) [reply pkg/controller/atlasdeployment/serverless_private_endpoint.go:205]: or at least add a godoc comment why `GetComment` is associated with the name.

File: pkg/api/v1/status/serverless_private_endpoint.go [hunk 3]
```diff
@@ -27,48 +25,16 @@ type ServerlessPrivateEndpoint struct {
...
-		CloudProviderEndpointID:      pe.CloudProviderEndpointID,
-		Name:                         pe.Comment,
-		EndpointServiceName:          pe.EndpointServiceName,
-		ErrorMessage:                 message,
-		Status:                       StatusFailed,
-		ProviderName:                 pe.ProviderName,
-		PrivateEndpointIPAddress:     pe.PrivateEndpointIPAddress,
-		PrivateLinkServiceResourceID: pe.PrivateLinkServiceResourceID,
+		ID:                           in.GetId(),
+		Name:                         in.GetComment(),
```
Comment (s-urbaniak) [inline pkg/api/v1/status/serverless_private_endpoint.go:31]: This seems counter-intuitive. Please make a godoc comment why `GetComment()` is associated with the name.

File: pkg/controller/atlasdeployment/serverless_private_endpoint.go [hunk 4]
```diff
@@ -1,459 +1,271 @@
...
 	}
 
-	return &result
+	return toCreate, toUpdate, toDelete
 }
 
-func isReadySPEEqual(existingPE mongodbatlas.ServerlessPrivateEndpointConnection, desiredPE akov2.ServerlessPrivateEndpoint) bool {
-	return existingPE.Comment == desiredPE.Name && desiredPE.CloudProviderEndpointID == existingPE.CloudProviderEndpointID && desiredPE.PrivateEndpointIPAddress == existingPE.PrivateEndpointIPAddress
-}
+func mapStatuses(atlasPrivateEndpoints []admin.ServerlessTenantEndpoint) map[string]status.ServerlessPrivateEndpoint {
```
Comment (s-urbaniak) [inline pkg/controller/atlasdeployment/serverless_private_endpoint.go:237]: rename this to `statusMap`, also note that plural of "status" is "status", see https://www.grammarbook.com/blog/singular-vs-plural/plural-of-status

File: pkg/controller/atlasdeployment/serverless_private_endpoint.go [hunk 5]
```diff
@@ -1,459 +1,271 @@
...
-	logger := service.Log
-	prevCfg := prevPEConfig(latestConfig)
-	if matchingPEs(logger, deployment.Spec.ServerlessSpec.PrivateEndpoints, existingPE) ||
-		matchingPEs(logger, prevCfg, existingPE) {
-		return true, nil
-	}
-	return false, nil
-}
+	toCreate, toUpdate, toDelete := sortTasks(deployment.PrivateEndpoints, atlasPrivateEndpoints)
+	speStatusesMap := mapStatuses(atlasPrivateEndpoints)
```
Comment (s-urbaniak) [inline pkg/controller/atlasdeployment/serverless_private_endpoint.go:68]: s/speStatusesMap/speStatusMap

File: pkg/controller/atlasdeployment/serverless_private_endpoint.go [hunk 6]
```diff
@@ -1,459 +1,271 @@
...
 
-	return &result
+	return false
 }
 
-func getAllExistingServerlessPE(ctx context.Context, service mongodbatlas.ServerlessPrivateEndpointsService, groupID, clusterName string) ([]mongodbatlas.ServerlessPrivateEndpointConnection, error) {
-	list, _, err := service.List(ctx, groupID, clusterName, nil)
-	if err != nil {
-		return nil, fmt.Errorf("failed to list serverless private endpoints: %w", err)
+func checkStatuses(pe []status.ServerlessPrivateEndpoint) bool {
```
Comment (s-urbaniak) [inline pkg/controller/atlasdeployment/serverless_private_endpoint.go:263]: s/checkStatuses/hasUnavailable

File: pkg/controller/atlasdeployment/serverless_private_endpoint_test.go [hunk 7]
```diff
@@ -2,289 +2,576 @@ package atlasdeployment
...
+		}
+		result := ensureServerlessPrivateEndpoints(&service, "project-id", &deployment)
 
-		require.NoError(t, err)
-		assert.False(t, result)
+		assert.Equal(
+			t,
+			workflow.InProgress(workflow.ServerlessPrivateEndpointInProgress, "Waiting serverless private endpoint to be configured"),
+			result,
+		)
```
Comment (s-urbaniak) [inline pkg/controller/atlasdeployment/serverless_private_endpoint_test.go:199]: add the following assertions:
```
		c, ok := service.GetCondition(status.ServerlessPrivateEndpointReadyType)
		assert.True(t, ok)
		c.LastTransitionTime = metav1.Time{} // not relevant for the unit test
		assert.Equal(t, status.Condition{
			Type:    status.ServerlessPrivateEndpointReadyType,
			Status:  corev1.ConditionFalse,
			Reason:  string(workflow.ServerlessPrivateEndpointInProgress),
			Message: "Waiting serverless private endpoint to be configured",
		}, c)
```

**Pull request merged**